src/workflow.ts

```diff
@@ -0,0 +1,160 @@
+import * as proc from '@cmt/proc';
+import { ProjectController } from './projectController';
+
+nls.config({ messageFormat: nls.MessageFormat.bundle, bundleFormat: nls.BundleFormat.standalone })();
+const localize: nls.LocalizeFunc = nls.loadMessageBundle();
+
+import * as logging from './logging';
+import { ConfigureTrigger } from './cmakeProject';
+const log = logging.createLogger('workflow');
+
+// class WorkflowOutputLogger implements OutputConsumer {
+//     output(line: string) {
+//         log.info(line);
+//     }
+//     error(line: string) {
+//         this.output(line);
+//     }
+// }
+
+export class WorkflowDriver implements vscode.Disposable {
+    /**
+     * @param projectController
+     */
+    constructor(readonly ws: DirectoryContext, private readonly projectController?: ProjectController) {}
+
+    dispose() {
+    }
+
+    public async runWorkflow(driver: CMakeDriver,
+        workflowPreset?: WorkflowPreset | null,
+        configurePreset?: ConfigurePreset | null,
+        buildPreset?: BuildPreset | null,
+        testPreset?: TestPreset | null,
+        packagePreset?: PackagePreset | null,
+        _consumer?: proc.OutputConsumer): Promise<number> {
+        if (driver.useCMakePresets && !driver.workflowPreset) {
+            log.error(localize('workflow.preset.not.set', 'Workflow preset is not set'));
+            return -3;
+        }
+        const oldConfigurePreset = configurePreset;
+        const oldBuildPreset = buildPreset;
+        const oldTestPreset = testPreset;
+        const oldPackagePreset = packagePreset;
+
+        if (workflowPreset?.steps[0].type !== "configure") {
+            log.error(localize('workflow.does.not.start.configure.step', 'The workflow does not start with a configure step.'));
+            return -3;
+        }
+
+        const prj = await this.projectController?.getProjectForFolder(driver.workspaceFolder);
+        if (!prj) {
+            log.error(localize('no.project.found', 'No project found for folder {0}', driver.workspaceFolder));
+            return -2;
+        }
+
+        let newConfigurePreset: ConfigurePreset | null = null;
+        let newBuildPreset: BuildPreset | null = null;
+        let newTestPreset: TestPreset | null = null ;
+        let newPackagePreset: PackagePreset | null = null;
+        for (const step of workflowPreset.steps) {
+            switch (step.type) {
+                case "configure":
+                    newConfigurePreset = getPresetByName(allConfigurePresets(driver.workspaceFolder), step.name);
+                    if (!newConfigurePreset) {
+                        log.error(localize('workflow.configure.preset.does.not.exist', 'The workflow step references a non existing configure preset: {0}', step.name));
+                        return -3;
+                    }
+
+                    if (step.name !== workflowPreset.steps[0].name) {
```

Comment:
Do we need this both here and above the switch statement? Technically it's double protection, but if we don't need it in both places, we should remove one (probably the one above the switch)